# Load File

In [37]:
# Imports

import pandas as pd 
import numpy as np

In [38]:
# Load CSV file 

df = pd.read_csv('V:/Project/Project_Stick/src/tnrera_projects_for_modelling.csv')


In [39]:
pd.set_option('display.max_columns', None)
df.head()

,Project_Name,Project_Registration_No,Project_Details,Districts,Latitude,Longitude,Usage,Type_of_Building,Site_Extent_Sqm,FSI_LIG_Residential,FSI_Other_Residential,FSI_Commercial,Total_Dwelling_Units,Land_Cost,Construction_Cost,Other_Cost,Total_Project_Cost,Covered_Parking,Open_Parking,Visitor_Parking,Floor_mentions,Heights_m,FE_Site_Extent_Sqft,FE_Total_FSI_Area_Sqft,FE_FSI,FE_Permissable_FSI,FE_Max_Premium_FSI,FE_Total_Parking,FE_Land_Cost_per_Sqft,FE_Construction_Cost_per_Sqft,FE_Total_Project_Cost_per_Sqft,FE_Land_Cost_Ratio,FE_Construction_Cost_Ratio,FE_Cost_Per_Dwelling_Unit,FE_Average_Unit_Size_Sqft,Is_TNUHDB_TNHB,Is_Dwelling_Units_Hardcoded,Is_FSI_Imputed_From_Comparable_Avg,Coord_Was_Corrupted,Is_Coordinates_Hardcoded,District,Usage_clean,FSI_outlier_flag,Imputed_Car_Parking,Parking_was_imputed,Is_Construction_Cost_Below_Floor,Is_Construction_Cost_Imputed
0,CRT Magilagam,TN/10/Building/0519/2024,"Ward-E,Block-21,TS No.2/3,Periyar Nagar,Erode ...",Erode,11.335912,77.725072,Residential,Non-High Rise Building (NHRB),1149,0.0,3150.0,NaN,30.0,24731214,111271043,16491620.0,152493877,22.0,6.0,3.0,[],[],12370,33906,2.74,2.0,2.00,31.0,1999,3282.0,4498,0.16,0.73,5083129.0,1130.0,False,False,False,0,False,Erode,Residential,False,NaN,False,False,False
1,Mayflower East Gate,TN/11/Building/0301/2024,"Old S.F.No.418/2pt,New T.S.No.5/1,Block.07,War...",Coimbatore,11.219200,77.113400,Residential,High Rise Building (HRB),2230,NaN,6260.0,NaN,42.0,80000000,230000000,15982769.0,325982769,27.0,15.0,42.0,[],[],24003,67383,2.81,2.0,4.87,84.0,3333,3413.0,4838,0.25,0.71,7761494.0,1604.0,False,False,False,0,False,Coimbatore,Residential,False,NaN,False,False,False
2,PNR MODERNA,TN/11/Building/0344/2024,"SF NO 129/4B, RANI GARDEN, 3RD STREET,SINGANAL...",Coimbatore,11.006265,77.042583,Residential,Non-High Rise Building (NHRB),720,NaN,1818.0,NaN,14.0,26016000,43747218,11130881.0,80894099,16.0,0.0,0.0,[],[],7750,19572,2.53,2.0,2.00,16.0,3357,2235.0,4133,0.32,0.54,5778150.0,1398.0,False,False,False,0,False,Coimbatore,Residential,False,NaN,False,False,False
3,SHRESTHA VILLAS https://rera.tn.gov.in/public-...,TN/11/Building/0346/2024,Block1: Ground Floor + 1st Floor with 4 Villas...,Coimbatore,10.511000,10.574000,Residential,Non-High Rise Building (NHRB),81550,NaN,16100.0,NaN,400.0,1200000000,0,0.0,1200000000,246.0,0.0,0.0,"[{'block': 'Block 2', 'base_level': 'Ground', ...",[],877796,173299,0.20,2.0,2.00,246.0,1367,3307.0,6924,1.00,0.00,3000000.0,433.0,False,False,False,0,False,Coimbatore,Residential,False,NaN,False,True,True
4,ADHYA,TN/11/Building/0353/2024,"Adhya, SF.NO:577/1B2A, Chinnavedampatti Villag...",Coimbatore,11.043040,76.590390,Residential,Non-High Rise Building (NHRB),5047,NaN,14025.0,NaN,200.0,54450000,835820000,3000000.0,893270000,95.0,0.0,13.0,[],[],54324,150965,2.78,2.0,2.00,108.0,1002,5537.0,5917,0.06,0.94,4466350.0,755.0,False,False,False,0,False,Coimbatore,Residential,False,NaN,False,False,False


In [40]:
df.shape

(943, 47)

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 47 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Project_Name                        943 non-null    object 
 1   Project_Registration_No             943 non-null    object 
 2   Project_Details                     943 non-null    object 
 3   Districts                           943 non-null    object 
 4   Latitude                            943 non-null    float64
 5   Longitude                           943 non-null    float64
 6   Usage                               943 non-null    object 
 7   Type_of_Building                    943 non-null    object 
 8   Site_Extent_Sqm                     943 non-null    int64  
 9   FSI_LIG_Residential                 828 non-null    float64
 10  FSI_Other_Residential               928 non-null    float64
 11  FSI_Commercial                      338 non-n

# Feature Engineering

In [42]:
# For Modelling without FE_FSI_Headroom

# df_set1 — For HGB (handle NaN + categorical dtype natively)
df_set1 = df.copy()

# df_set2 — For LightGBM (handle NaN + categorical dtype natively)
df_set2 = df.copy()

# df_set3 — For XGBoost (handle NaN + categorical dtype natively)
df_set3 = df.copy()


print(f"df_set1: {df_set1.shape}")
print(f"df_set2: {df_set2.shape}")
print(f"df_set3: {df_set3.shape}")



df_set1: (943, 47)
df_set2: (943, 47)
df_set3: (943, 47)


# Modelling 

## Gradient Boosting Decision Tree (GBDT)

### Drop Columns

Drop Project_Name, Project_Registration_No,	Project_Details, Floor_mentions, Heights_m, District, Usage_clean

In [43]:
df_set1.columns

Index(['Project_Name', 'Project_Registration_No', 'Project_Details',
       'Districts', 'Latitude', 'Longitude', 'Usage', 'Type_of_Building',
       'Site_Extent_Sqm', 'FSI_LIG_Residential', 'FSI_Other_Residential',
       'FSI_Commercial', 'Total_Dwelling_Units', 'Land_Cost',
       'Construction_Cost', 'Other_Cost', 'Total_Project_Cost',
       'Covered_Parking', 'Open_Parking', 'Visitor_Parking', 'Floor_mentions',
       'Heights_m', 'FE_Site_Extent_Sqft', 'FE_Total_FSI_Area_Sqft', 'FE_FSI',
       'FE_Permissable_FSI', 'FE_Max_Premium_FSI', 'FE_Total_Parking',
       'FE_Land_Cost_per_Sqft', 'FE_Construction_Cost_per_Sqft',
       'FE_Total_Project_Cost_per_Sqft', 'FE_Land_Cost_Ratio',
       'FE_Construction_Cost_Ratio', 'FE_Cost_Per_Dwelling_Unit',
       'FE_Average_Unit_Size_Sqft', 'Is_TNUHDB_TNHB',
       'Is_Dwelling_Units_Hardcoded', 'Is_FSI_Imputed_From_Comparable_Avg',
       'Coord_Was_Corrupted', 'Is_Coordinates_Hardcoded', 'District',
       'Usage_clean', 'FSI_outlier

In [44]:
# Drop columns that are mentioned above. We are dropping these because they have no bearing on modelling

df_set1 = df_set1.drop(['Project_Name', 'Project_Registration_No', 'Project_Details',
               'Floor_mentions', 'Heights_m', 'District', 'Usage_clean',], axis=1)


In [45]:
# Drop Rows which have 'Is_TNUHDB_TNHB'==TRUE


drop_index = df_set1[df_set1['Is_TNUHDB_TNHB'] == True].index
print("No of Rows being dropped :",len(drop_index))
df_set1 = df_set1.drop(drop_index)
print("No of Rows in df_set1 :",len(df_set1))

No of Rows being dropped : 0
No of Rows in df_set1 : 943


### Model

In [46]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import HistGradientBoostingRegressor
from sklearn.metrics import r2_score, mean_absolute_error, median_absolute_error
import numpy as np

# ---------------------------------------------------------------------------
# TARGET — log-scale, per your earlier modeling decision:
# raw Total_Project_Cost_per_Sqft is right-skewed, log stabilizes variance
# ---------------------------------------------------------------------------
TARGET = 'FE_Total_Project_Cost_per_Sqft'

df_set1 = df_set1.dropna(subset=[TARGET]).copy()  # rows missing the target can't be used for training
y = np.log1p(df_set1[TARGET])

# ---------------------------------------------------------------------------
# FEATURES — everything except the target itself and any other cost-derived
# columns that would leak target information (e.g. FE_Construction_Cost_per_Sqft
# is a component of Total_Project_Cost, so including it lets the model cheat)
# ---------------------------------------------------------------------------
LEAKAGE_COLS = [
    TARGET, 'Total_Project_Cost', 'Land_Cost', 'Construction_Cost', 'Other_Cost',
    'FE_Land_Cost_per_Sqft', 'FE_Construction_Cost_per_Sqft',
    'FE_Land_Cost_Ratio', 'FE_Construction_Cost_Ratio','FE_Cost_Per_Dwelling_Unit'
]
X = df_set1.drop(columns=[c for c in LEAKAGE_COLS if c in df_set1.columns])

# Convert categorical columns to pandas 'category' dtype — HGB reads this natively
categorical_cols = ['Districts', 'Usage', 'Type_of_Building']
for col in categorical_cols:
    if col in X.columns:
        X[col] = X[col].astype('category')

print(f"Rows: {len(X)}  |  Features: {X.shape[1]}")
print(f"Feature columns: {list(X.columns)}")

# ---------------------------------------------------------------------------
# TRAIN/TEST SPLIT
# ---------------------------------------------------------------------------
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# ---------------------------------------------------------------------------
# FIT — categorical_features='from_dtype' tells HGB to auto-detect the
# category-dtype columns rather than requiring explicit column indices
# ---------------------------------------------------------------------------
hgb = HistGradientBoostingRegressor(categorical_features='from_dtype', random_state=42)
hgb.fit(X_train, y_train)

# ---------------------------------------------------------------------------
# EVALUATE — back-transform from log scale before computing error metrics,
# so R²/MAE are reported in real rupee/sqft terms, not log-space
# ---------------------------------------------------------------------------
y_pred_log = hgb.predict(X_test)
y_pred = np.expm1(y_pred_log)
y_test_actual = np.expm1(y_test)

r2 = r2_score(y_test, y_pred_log)  # R² computed on log scale — matches what was actually optimized
mae = mean_absolute_error(y_test_actual, y_pred)
median_ape = np.median(np.abs((y_test_actual - y_pred) / y_test_actual)) * 100

print(f"\nHGB Results:")
print(f"R² (log scale): {r2:.3f}")
print(f"MAE (rupees/sqft): {mae:,.0f}")
print(f"Median APE: {median_ape:.1f}%")

Rows: 943  |  Features: 30
Feature columns: ['Districts', 'Latitude', 'Longitude', 'Usage', 'Type_of_Building', 'Site_Extent_Sqm', 'FSI_LIG_Residential', 'FSI_Other_Residential', 'FSI_Commercial', 'Total_Dwelling_Units', 'Covered_Parking', 'Open_Parking', 'Visitor_Parking', 'FE_Site_Extent_Sqft', 'FE_Total_FSI_Area_Sqft', 'FE_FSI', 'FE_Permissable_FSI', 'FE_Max_Premium_FSI', 'FE_Total_Parking', 'FE_Average_Unit_Size_Sqft', 'Is_TNUHDB_TNHB', 'Is_Dwelling_Units_Hardcoded', 'Is_FSI_Imputed_From_Comparable_Avg', 'Coord_Was_Corrupted', 'Is_Coordinates_Hardcoded', 'FSI_outlier_flag', 'Imputed_Car_Parking', 'Parking_was_imputed', 'Is_Construction_Cost_Below_Floor', 'Is_Construction_Cost_Imputed']

HGB Results:
R² (log scale): 0.299
MAE (rupees/sqft): 2,080
Median APE: 27.5%


In [47]:
# Difference in actual and predicted - error

residuals_df = pd.DataFrame({
    'actual': y_test_actual,
    'predicted': y_pred,
    'abs_pct_error': np.abs((y_test_actual - y_pred) / y_test_actual) * 100
}).sort_values('abs_pct_error', ascending=False)

print(residuals_df.head(15))

      actual     predicted  abs_pct_error
506   2019.0   8600.366100     325.971575
109   3391.0   8940.384302     163.650378
309   3149.0   7808.252776     147.959758
559   2727.0   6410.226661     135.065151
611   2848.0   6312.411737     121.643671
139   5025.0  10437.628410     107.713998
468   4407.0   9075.006045     105.922533
778   2963.0   5901.300543      99.166404
44    2452.0   4809.153186      96.131859
449   2090.0   4070.523043      94.761868
497   4542.0   8718.945208      91.962686
925   3264.0   6237.803542      91.109177
310   6101.0  11579.793268      89.801562
884  24213.0   4680.762210      80.668392
477   2817.0   4955.700921      75.921225


In [48]:
# Checking permutation importance

from sklearn.inspection import permutation_importance

result = permutation_importance(hgb, X_test, y_test, n_repeats=10, random_state=42, n_jobs=-1)

importance_df = pd.DataFrame({
    'feature': X_test.columns,
    'importance_mean': result.importances_mean,
    'importance_std': result.importances_std
}).sort_values('importance_mean', ascending=False)

print(importance_df.to_string(index=False))

                           feature  importance_mean  importance_std
                         Districts         0.186552        0.067196
  Is_Construction_Cost_Below_Floor         0.062914        0.011241
                         Longitude         0.051657        0.021868
                          Latitude         0.046677        0.033046
         FE_Average_Unit_Size_Sqft         0.038200        0.021515
               FSI_LIG_Residential         0.010660        0.007717
                   Covered_Parking         0.007916        0.012299
                   Visitor_Parking         0.006135        0.010053
                   Site_Extent_Sqm         0.006106        0.017734
               Coord_Was_Corrupted         0.004112        0.001559
                  FE_Total_Parking         0.001035        0.005733
                  Type_of_Building         0.000701        0.001992
                             Usage         0.000560        0.001360
                    Is_TNUHDB_TNHB         0.000

### Inference 



From 943 rows and 30 features, the model - HGB Results:
R² (log scale): 0.299
MAE (rupees/sqft): 2,080
Median APE: 27.5%


R² of 0.299 means the model explains only ~30% of the variance in log-price — the majority of price variation remains unexplained by the features available. This is not primarily a modeling failure; it reflects a real ceiling in the underlying TNRERA data: the filings do not capture finish quality, amenities, or micro-location factors that materially separate a luxury project from a budget one at the same coordinates. Two projects sitting a few hundred meters apart can have very different price points for reasons the dataset simply does not record.

A Mean Absolute Error of Rs 2,080 per sqft means that, on average, the model's predicted price is off from the actual price by Rs 2,080 per sqft — in either direction. So if a project actually sells for Rs 7,000/sqft, the model's prediction could land anywhere around Rs 4,920 to Rs 9,080/sqft on a typical miss.

This number alone doesn't tell you if the error is "bad" — it depends on the price range. At Rs 5,000/sqft, an average miss of Rs 2,080 is over 40% off; at Rs 15,000/sqft, the same miss is under 14%. That's exactly why the Median APE below matters more than MAE on its own — it normalizes the error against the actual price, so it's comparable across cheap and expensive projects. 

A Median Absolute Percentage Error of 27.5% indicates for  typical project,the models prediction is off my more that a 27.5% of the actual total project cost. E.g If a projects actual cost was Rs8000/sqft, the prediction can be anywhere between Rs5800/sqft to Rs10,200/sqft and still fall within a typical error band. This is a wide margin for a pricing tool meant to inform real decisions.


## Light Gradient Boost Model


### Drop Columns

In [49]:
## Drop Columns Project_Name, Project_Registration_No,	Project_Details, Floor_mentions, Heights_m, District, Usage_clean

df_set2.columns

Index(['Project_Name', 'Project_Registration_No', 'Project_Details',
       'Districts', 'Latitude', 'Longitude', 'Usage', 'Type_of_Building',
       'Site_Extent_Sqm', 'FSI_LIG_Residential', 'FSI_Other_Residential',
       'FSI_Commercial', 'Total_Dwelling_Units', 'Land_Cost',
       'Construction_Cost', 'Other_Cost', 'Total_Project_Cost',
       'Covered_Parking', 'Open_Parking', 'Visitor_Parking', 'Floor_mentions',
       'Heights_m', 'FE_Site_Extent_Sqft', 'FE_Total_FSI_Area_Sqft', 'FE_FSI',
       'FE_Permissable_FSI', 'FE_Max_Premium_FSI', 'FE_Total_Parking',
       'FE_Land_Cost_per_Sqft', 'FE_Construction_Cost_per_Sqft',
       'FE_Total_Project_Cost_per_Sqft', 'FE_Land_Cost_Ratio',
       'FE_Construction_Cost_Ratio', 'FE_Cost_Per_Dwelling_Unit',
       'FE_Average_Unit_Size_Sqft', 'Is_TNUHDB_TNHB',
       'Is_Dwelling_Units_Hardcoded', 'Is_FSI_Imputed_From_Comparable_Avg',
       'Coord_Was_Corrupted', 'Is_Coordinates_Hardcoded', 'District',
       'Usage_clean', 'FSI_outlier

In [50]:

# Drop columns that are mentioned above. We are dropping these because they have no bearing on modelling

df_set2 = df_set2.drop(['Project_Name', 'Project_Registration_No', 'Project_Details',
               'Floor_mentions', 'Heights_m', 'District', 'Usage_clean',], axis=1)

In [51]:


# Drop Rows which have 'Is_TNUHDB_TNHB'==TRUE


drop_index = df_set2[df_set2['Is_TNUHDB_TNHB'] == True].index
print("No of Rows being dropped :",len(drop_index))
df_set2 = df_set2.drop(drop_index)
print("No of Rows in df_set1 :",len(df_set2))
df_set2['FE_Total_Project_Cost_per_Sqft'].describe()

No of Rows being dropped : 0
No of Rows in df_set1 : 943


count      943.000000
mean      6613.966066
std       3736.829480
min       2003.000000
25%       4164.500000
50%       5608.000000
75%       8186.000000
max      40643.000000
Name: FE_Total_Project_Cost_per_Sqft, dtype: float64

### Model

In [52]:
object_cols = X.select_dtypes(include='object').columns.tolist()
print(f"Object columns found: {object_cols}")

for col in object_cols:
    if col not in categorical_cols:
        X[col] = X[col].astype('category')

print(X.dtypes[X.dtypes == 'object'])  # should be empty now

Object columns found: ['Parking_was_imputed']
Series([], dtype: object)


In [53]:
print(X.dtypes[X.dtypes == 'object'])

Series([], dtype: object)


In [54]:
if 'Parking_was_imputed' not in X.columns:
    print("Column missing from X — check df_set1 and the feature-selection cell above")
else:
    X['Parking_was_imputed'] = X['Parking_was_imputed'].fillna(False).astype(bool)

In [55]:
print('Parking_was_imputed' in df_set2.columns)

True


In [56]:
X['Parking_was_imputed'] = X['Parking_was_imputed'].fillna(False).astype(bool)

In [57]:
print(X['Parking_was_imputed'].dtype)
print(X['Parking_was_imputed'].unique())

bool
[False]


In [58]:
import lightgbm as lgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error
import numpy as np

TARGET = 'FE_Total_Project_Cost_per_Sqft'

df_set2 = df_set2.dropna(subset=[TARGET]).copy()
y = np.log1p(df_set2[TARGET])

LEAKAGE_COLS = [
    TARGET, 'Total_Project_Cost', 'Land_Cost', 'Construction_Cost', 'Other_Cost',
    'FE_Land_Cost_per_Sqft', 'FE_Construction_Cost_per_Sqft',
    'FE_Land_Cost_Ratio', 'FE_Construction_Cost_Ratio', 'FE_Cost_Per_Dwelling_Unit'
]
X = df_set2.drop(columns=[c for c in LEAKAGE_COLS if c in df_set2.columns])

categorical_cols = ['Districts', 'Usage', 'Type_of_Building']
for col in categorical_cols:
    if col in X.columns:
        X[col] = X[col].astype('category')

# fix the object-dtype boolean column
X['Parking_was_imputed'] = X['Parking_was_imputed'].fillna(False).astype(bool)

print(X.dtypes[X.dtypes == 'object'])  # must print empty
print(f"Rows: {len(X)}  |  Features: {X.shape[1]}")

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

lgb_model = lgb.LGBMRegressor(random_state=42)
lgb_model.fit(X_train, y_train, categorical_feature=categorical_cols)

y_pred_log = lgb_model.predict(X_test)
y_pred = np.expm1(y_pred_log)
y_test_actual = np.expm1(y_test)

r2 = r2_score(y_test, y_pred_log)
mae = mean_absolute_error(y_test_actual, y_pred)
median_ape = np.median(np.abs((y_test_actual - y_pred) / y_test_actual)) * 100

print(f"\nLightGBM Results:")
print(f"R² (log scale): {r2:.3f}")
print(f"MAE (rupees/sqft): {mae:,.0f}")
print(f"Median APE: {median_ape:.1f}%")

Series([], dtype: object)
Rows: 943  |  Features: 30
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000212 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 2400
[LightGBM] [Info] Number of data points in the train set: 754, number of used features: 22
[LightGBM] [Info] Start training from score 8.676838
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No f

C:\Users\Asus\AppData\Local\Temp\ipykernel_3244\564082708.py:24: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  X['Parking_was_imputed'] = X['Parking_was_imputed'].fillna(False).astype(bool)


### Inference

 MAE drops slightly to Rs 2,059/sqft and Median APE improves from 27.5% to 24.4% — while R² stays essentially flat at 0.299. That pattern is the key thing to note: better error metrics without a corresponding jump in R² means LightGBM is handling outliers and non-linear interactions in price per sqft better than the other models, but it isn't uncovering new signal in the data. The ~30% variance ceiling is a data constraint (missing finish quality, amenities, micro-location detail — as documented), not a model-capacity constraint, and no amount of algorithm swapping moves it much further.

A Median APE of 24.4% is still a real margin — a project actually priced at Rs 8,000/sqft could reasonably be predicted anywhere from ~Rs 6,050 to Rs 9,950/sqft. That's tighter than the earlier model but still too wide for precise per-project valuation.


## XGBoost Model

### Drop Columns

In [59]:
#Drop Project_Name, Project_Registration_No,	Project_Details, Floor_mentions, Heights_m, District, Usage_clean

df_set3.columns

Index(['Project_Name', 'Project_Registration_No', 'Project_Details',
       'Districts', 'Latitude', 'Longitude', 'Usage', 'Type_of_Building',
       'Site_Extent_Sqm', 'FSI_LIG_Residential', 'FSI_Other_Residential',
       'FSI_Commercial', 'Total_Dwelling_Units', 'Land_Cost',
       'Construction_Cost', 'Other_Cost', 'Total_Project_Cost',
       'Covered_Parking', 'Open_Parking', 'Visitor_Parking', 'Floor_mentions',
       'Heights_m', 'FE_Site_Extent_Sqft', 'FE_Total_FSI_Area_Sqft', 'FE_FSI',
       'FE_Permissable_FSI', 'FE_Max_Premium_FSI', 'FE_Total_Parking',
       'FE_Land_Cost_per_Sqft', 'FE_Construction_Cost_per_Sqft',
       'FE_Total_Project_Cost_per_Sqft', 'FE_Land_Cost_Ratio',
       'FE_Construction_Cost_Ratio', 'FE_Cost_Per_Dwelling_Unit',
       'FE_Average_Unit_Size_Sqft', 'Is_TNUHDB_TNHB',
       'Is_Dwelling_Units_Hardcoded', 'Is_FSI_Imputed_From_Comparable_Avg',
       'Coord_Was_Corrupted', 'Is_Coordinates_Hardcoded', 'District',
       'Usage_clean', 'FSI_outlier

In [60]:

# Drop columns that are mentioned above. We are dropping these because they have no bearing on modelling

df_set3 = df_set3.drop(['Project_Name', 'Project_Registration_No', 'Project_Details',
               'Floor_mentions', 'Heights_m', 'District', 'Usage_clean',], axis=1)

In [61]:
# Drop Rows which have 'Is_TNUHDB_TNHB'==TRUE


drop_index = df_set3[df_set3['Is_TNUHDB_TNHB'] == True].index
print("No of Rows being dropped :",len(drop_index))
df_set3 = df_set3.drop(drop_index)
print("No of Rows in df_set1 :",len(df_set3))
df_set3['FE_Total_Project_Cost_per_Sqft'].describe()

No of Rows being dropped : 0
No of Rows in df_set1 : 943


count      943.000000
mean      6613.966066
std       3736.829480
min       2003.000000
25%       4164.500000
50%       5608.000000
75%       8186.000000
max      40643.000000
Name: FE_Total_Project_Cost_per_Sqft, dtype: float64

### Model

In [62]:
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error
import numpy as np

TARGET = 'FE_Total_Project_Cost_per_Sqft'

df_set2 = df_set2.dropna(subset=[TARGET]).copy()
y = np.log1p(df_set2[TARGET])

LEAKAGE_COLS = [
    TARGET, 'Total_Project_Cost', 'Land_Cost', 'Construction_Cost', 'Other_Cost',
    'FE_Land_Cost_per_Sqft', 'FE_Construction_Cost_per_Sqft',
    'FE_Land_Cost_Ratio', 'FE_Construction_Cost_Ratio', 'FE_Cost_Per_Dwelling_Unit'
]
X = df_set2.drop(columns=[c for c in LEAKAGE_COLS if c in df_set2.columns])

categorical_cols = ['Districts', 'Usage', 'Type_of_Building']
for col in categorical_cols:
    if col in X.columns:
        X[col] = X[col].astype('category')

X['Parking_was_imputed'] = X['Parking_was_imputed'].fillna(False).astype(bool)

print(f"Rows: {len(X)}  |  Features: {X.shape[1]}")

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

xgb_model = xgb.XGBRegressor(random_state=42, enable_categorical=True)
xgb_model.fit(X_train, y_train)

y_pred_log = xgb_model.predict(X_test)
y_pred = np.expm1(y_pred_log)
y_test_actual = np.expm1(y_test)

r2 = r2_score(y_test, y_pred_log)
mae = mean_absolute_error(y_test_actual, y_pred)
median_ape = np.median(np.abs((y_test_actual - y_pred) / y_test_actual)) * 100

print(f"\nXGBoost Results:")
print(f"R² (log scale): {r2:.3f}")
print(f"MAE (rupees/sqft): {mae:,.0f}")
print(f"Median APE: {median_ape:.1f}%")

Rows: 943  |  Features: 30


C:\Users\Asus\AppData\Local\Temp\ipykernel_3244\3173959128.py:23: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  X['Parking_was_imputed'] = X['Parking_was_imputed'].fillna(False).astype(bool)



XGBoost Results:
R² (log scale): 0.220
MAE (rupees/sqft): 2,173
Median APE: 29.1%


In [63]:
from sklearn.inspection import permutation_importance
import pandas as pd

# ---------------------------------------------------------------------------
# Actual vs predicted with % error, sorted worst-first
# ---------------------------------------------------------------------------
results = pd.DataFrame({
    'actual': y_test_actual,
    'predicted': y_pred,
})
results['abs_pct_error'] = np.abs((results['actual'] - results['predicted']) / results['actual']) * 100
results = results.sort_values('abs_pct_error', ascending=False)

print(results.head(15))

# ---------------------------------------------------------------------------
# Permutation importance — computed on the held-out test set, log scale
# ---------------------------------------------------------------------------
perm = permutation_importance(xgb_model, X_test, y_test, n_repeats=10, random_state=42, n_jobs=-1)

importance_df = pd.DataFrame({
    'feature': X_test.columns,
    'importance_mean': perm.importances_mean,
    'importance_std': perm.importances_std
}).sort_values('importance_mean', ascending=False)

print(importance_df)

     actual     predicted  abs_pct_error
506  2019.0   7982.873047     295.387471
309  3149.0   8778.731445     178.778388
109  3391.0   9363.748047     176.135301
266  2003.0   4527.837402     126.052791
559  2727.0   6049.708008     121.844811
925  3264.0   6912.005859     111.764885
139  5025.0  10377.145508     106.510358
497  4542.0   9216.046875     102.907241
67   2760.0   5568.996582     101.775238
355  2963.0   5936.865723     100.366714
449  2090.0   4050.145996      93.786890
451  4096.0   7915.825684      93.257463
611  2848.0   5447.216309      91.264618
254  2913.0   5369.250000      84.320288
703  4931.0   9087.734375      84.298000
                               feature  importance_mean  importance_std
0                            Districts         0.346330        0.081582
2                            Longitude         0.096200        0.033251
14              FE_Total_FSI_Area_Sqft         0.080644        0.035682
1                             Latitude         0.064812 

### Inference

R² drops to 0.220 (down from 0.299), MAE rises to Rs 2,173/sqft, and Median APE worsens to 29.1%. All three metrics move in the same direction, which is different from the LightGBM case: this isn't "same ceiling, better error handling," it's genuinely explaining less of the variance and predicting less accurately across the board.

A Median APE of 29.1% means a project actually priced at Rs 8,000/sqft could be predicted anywhere from ~Rs 5,670 to Rs 10,330/sqft on a typical miss — the widest margin of the three models tested.



# Model Comparison

| Model | R² (log scale) | MAE (₹/sqft) | Median APE | ₹8,000/sqft prediction range |
|---|---|---|---|---|
| Linear Regression (baseline) | 0.299 | 2,080 | 27.5% | ₹5,800 – ₹10,200 |
| LightGBM | 0.299 | 2,059 | 24.4% | ₹6,050 – ₹9,950 |
| XGBoost | 0.220 | 2,173 | 29.1% | ₹5,670 – ₹10,330 |

These models are useful for identifying relative positioning — is a project priced high or low for its district/typology — but it should not be treated as a precise per-project valuation tool. 

# Output for Streamlit Dashboard

In [64]:
import joblib

joblib.dump(lgb_model, 'lgbm_avm_model.pkl')

feature_columns = X.columns.tolist()
joblib.dump(feature_columns, 'lgbm_avm_features.pkl')

joblib.dump(categorical_cols, 'lgbm_avm_categorical_cols.pkl')

print(f"Model saved. Trained on {len(feature_columns)} features (log1p target).")
print(feature_columns)

Model saved. Trained on 30 features (log1p target).
['Districts', 'Latitude', 'Longitude', 'Usage', 'Type_of_Building', 'Site_Extent_Sqm', 'FSI_LIG_Residential', 'FSI_Other_Residential', 'FSI_Commercial', 'Total_Dwelling_Units', 'Covered_Parking', 'Open_Parking', 'Visitor_Parking', 'FE_Site_Extent_Sqft', 'FE_Total_FSI_Area_Sqft', 'FE_FSI', 'FE_Permissable_FSI', 'FE_Max_Premium_FSI', 'FE_Total_Parking', 'FE_Average_Unit_Size_Sqft', 'Is_TNUHDB_TNHB', 'Is_Dwelling_Units_Hardcoded', 'Is_FSI_Imputed_From_Comparable_Avg', 'Coord_Was_Corrupted', 'Is_Coordinates_Hardcoded', 'FSI_outlier_flag', 'Imputed_Car_Parking', 'Parking_was_imputed', 'Is_Construction_Cost_Below_Floor', 'Is_Construction_Cost_Imputed']


In [65]:
df_set2.head(10)

,Districts,Latitude,Longitude,Usage,Type_of_Building,Site_Extent_Sqm,FSI_LIG_Residential,FSI_Other_Residential,FSI_Commercial,Total_Dwelling_Units,Land_Cost,Construction_Cost,Other_Cost,Total_Project_Cost,Covered_Parking,Open_Parking,Visitor_Parking,FE_Site_Extent_Sqft,FE_Total_FSI_Area_Sqft,FE_FSI,FE_Permissable_FSI,FE_Max_Premium_FSI,FE_Total_Parking,FE_Land_Cost_per_Sqft,FE_Construction_Cost_per_Sqft,FE_Total_Project_Cost_per_Sqft,FE_Land_Cost_Ratio,FE_Construction_Cost_Ratio,FE_Cost_Per_Dwelling_Unit,FE_Average_Unit_Size_Sqft,Is_TNUHDB_TNHB,Is_Dwelling_Units_Hardcoded,Is_FSI_Imputed_From_Comparable_Avg,Coord_Was_Corrupted,Is_Coordinates_Hardcoded,FSI_outlier_flag,Imputed_Car_Parking,Parking_was_imputed,Is_Construction_Cost_Below_Floor,Is_Construction_Cost_Imputed
0,Erode,11.335912,77.725072,Residential,Non-High Rise Building (NHRB),1149,0.0,3150.0,NaN,30.0,24731214,111271043,16491620.0,152493877,22.0,6.0,3.0,12370,33906,2.74,2.0,2.00,31.0,1999,3282.0,4498,0.16,0.73,5083129.0,1130.0,False,False,False,0,False,False,NaN,False,False,False
1,Coimbatore,11.219200,77.113400,Residential,High Rise Building (HRB),2230,NaN,6260.0,NaN,42.0,80000000,230000000,15982769.0,325982769,27.0,15.0,42.0,24003,67383,2.81,2.0,4.87,84.0,3333,3413.0,4838,0.25,0.71,7761494.0,1604.0,False,False,False,0,False,False,NaN,False,False,False
2,Coimbatore,11.006265,77.042583,Residential,Non-High Rise Building (NHRB),720,NaN,1818.0,NaN,14.0,26016000,43747218,11130881.0,80894099,16.0,0.0,0.0,7750,19572,2.53,2.0,2.00,16.0,3357,2235.0,4133,0.32,0.54,5778150.0,1398.0,False,False,False,0,False,False,NaN,False,False,False
3,Coimbatore,10.511000,10.574000,Residential,Non-High Rise Building (NHRB),81550,NaN,16100.0,NaN,400.0,1200000000,0,0.0,1200000000,246.0,0.0,0.0,877796,173299,0.20,2.0,2.00,246.0,1367,3307.0,6924,1.00,0.00,3000000.0,433.0,False,False,False,0,False,False,NaN,False,True,True
4,Coimbatore,11.043040,76.590390,Residential,Non-High Rise Building (NHRB),5047,NaN,14025.0,NaN,200.0,54450000,835820000,3000000.0,893270000,95.0,0.0,13.0,54324,150965,2.78,2.0,2.00,108.0,1002,5537.0,5917,0.06,0.94,4466350.0,755.0,False,False,False,0,False,False,NaN,False,False,False
5,Coimbatore,11.004030,77.024250,Residential,Non-High Rise Building (NHRB),2349,NaN,2022.0,NaN,18.0,75851295,60624219,12000000.0,148475514,0.0,19.0,3.0,25284,21768,0.86,2.0,2.00,22.0,3000,2785.0,6821,0.51,0.41,8248640.0,1209.0,False,False,False,0,False,False,NaN,False,False,False
6,Coimbatore,11.004850,77.013800,Residential,Non-High Rise Building (NHRB),10168,NaN,5853.0,404.0,100.0,120000000,152500000,44642575.0,317142575,64.0,12.0,0.0,109443,67349,0.62,2.0,2.00,76.0,1096,2264.0,4709,0.38,0.48,3171426.0,630.0,False,False,False,1,False,False,NaN,False,False,False
7,Coimbatore,11.054526,76.899453,Residential,High Rise Building (HRB),36418,NaN,22947.0,3460.0,89.0,301200000,1700000000,100000000.0,2101200000,178.0,143.0,22.0,392004,284238,0.73,2.0,4.87,343.0,768,5981.0,7392,0.14,0.81,23608989.0,2775.0,False,False,False,0,False,False,NaN,False,False,False
8,Coimbatore,11.079439,76.977709,Residential,Non-High Rise Building (NHRB),8675,NaN,17617.0,NaN,190.0,200000000,519742460,23128090.0,742870550,147.0,1.0,0.0,93377,189626,2.03,2.0,2.00,148.0,2142,2741.0,3918,0.27,0.70,3909845.0,998.0,False,False,False,0,False,False,NaN,False,False,False
9,Coimbatore,11.015088,76.950905,Commercial,Non-High Rise Building (NHRB),928,NaN,NaN,2592.0,NaN,34961500,78324000,7832400.0,121117900,22.0,0.0,2.0,9989,27900,2.79,2.0,2.00,24.0,3500,2807.0,4341,0.29,0.65,NaN,NaN,False,False,False,0,False,False,NaN,False,False,False
